libs/store_helper/ebpf_store_helper.c

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "ebpf_program_types.h"
+#include "ebpf_registry_helper.h"
+#include "ebpf_store_helper.h"
+#include "ebpf_windows.h"
+
+#define IS_SUCCESS(x) (x == EBPF_SUCCESS)
+
+static ebpf_result_t
+_ebpf_store_open_or_create_provider_registry_key(_Out_ ebpf_store_key_t* provider_key)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    ebpf_store_key_t root_key = NULL;
+    *provider_key = NULL;
+
+    // Open (or create) root eBPF registry path.
+#ifdef USER_MODE
```

Comment:
suggestion define `ebpf_store_root_key` differently for user mode and kernel mode, then you dont have to differentiate in the `ebpf_create_registry_key` call.

Reply:
The "sub-key" would also need to be part of the differentiation as well.

I refactored all the existing so that the `ebpf_store_root_key` and the new `ebpf_store_root_sub_key` variables are now atomically defined in their respective KM/UM `ebpf_registry_helper` libraries (and not in the projects that uses them, which was not ideal), this also further simplifies the code that uses the libraries.

Reply:
I'd asked why they're different in the first place, and I still haven't heard any answer to that.  Whatever the answer is, it should be documented in // comments at least, in the file(s).

Reply:
By looking to the code, it boils down to the different APIs used or KM/UM, i.e. `ZwCreateKey` and `RegCreateKeyExA` respectively. The first, takes a NULL (for OBJECT_ATTRIBUTES->RootDirectory) to use the given path as a fully qualified path, whereas the second builds the fully qualified path from predefined keys (https://learn.microsoft.com/en-us/windows/win32/sysinfo/predefined-keys):
- https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-zwcreatekey
- https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regcreatekeyexa

But I'd still call out @saxena-anurag for the actual history of this implementation.

Reply:
I think what Gianni mentioned was the reason to use #ifdef USER_MODE earlier -- to wrap 2 different registry APIs in a single common API (`ebpf_create_registry_key()`)

Reply:
ZwCreateKey exists in user-mode too, so is not a reason to have a difference.